app/services/subscriptions/dates_service.rb

```diff
@@ -0,0 +1,297 @@
+        next_end_of_weekly_period(date)
+      when :monthly
+        next_end_of_monthly_period(date)
+      when :yearly
+        next_end_of_yearly_period(date)
+      else
+        raise NotImplementedError
+      end
+    end
+
+    private
+
+    attr_accessor :subscription, :billing_date
+
+    delegate :plan, :subscription_date, :calendar?, to: :subscription
+
+    def base_date
+      @base_date ||= case plan.interval.to_sym
+                     when :weekly
+                       billing_date - 1.week
+                     when :monthly
+                       billing_date - 1.month
+                     when :yearly
+                       billing_date - 1.year
+                     else
+                       raise NotImplementedError
+      end
+    end
+
+    def subscription_day_name
+      subscription_date.strftime('%A').downcase.to_sym
+    end
+
+    def terminated_pay_in_arrear?
+      # NOTE: In case of termination or upgrade when we are terminating old plan (paying in arrear),
+      #       we should take to the beginning of the billing period
+      subscription.terminated? && plan.pay_in_arrear? && !subscription.downgraded?
+    end
+
+    def weekly_from_date
+      if terminated_pay_in_arrear?
+        return subscription.anniversary? ? billing_date.prev_occurring(subscription_day_name) : billing_date.beginning_of_week
+      end
+
+      subscription.anniversary? ? base_date.prev_occurring(subscription_day_name) : base_date.beginning_of_week
+    end
+
+    def weekly_to_date
+      from_date + 6.days
+    end
+
+    def monthly_from_date(date = base_date)
+      if terminated_pay_in_arrear?
+        return subscription.anniversary? ? previous_monthly_anniversary_day(billing_date) : billing_date.beginning_of_month
+      end
+
+      subscription.anniversary? ? previous_monthly_anniversary_day(date) : date.beginning_of_month
+    end
+
+    def monthly_to_date
+      return from_date.end_of_month if subscription.calendar? || subscription_date.day == 1
+
+      year = from_date.year
+      month = from_date.month + 1
+      day = subscription_date.day - 1
+
+      if month > 12
+        month = 1
+        year += 1
+      end
+
+      build_date(year, month, day)
+    end
+
+    def yearly_from_date
+      if terminated_pay_in_arrear?
+        return subscription.anniversary? ? previous_yearly_anniversary_day(billing_date) : billing_date.beginning_of_year
+      end
+
+      subscription.anniversary? ? previous_yearly_anniversary_day(base_date) : base_date.beginning_of_year
+    end
+
+    def yearly_to_date
+      return from_date.end_of_year if subscription.calendar? || subscription_date.day == 1
+
+      year = from_date.year + 1
+      month = from_date.month
+      day = subscription_date.day - 1
+
+      build_date(year, month, day)
+    end
+
+    def previous_monthly_anniversary_day(date)
+      year = nil
+      month = nil
+      day = subscription_date.day
+
+      if date.day <= subscription_date.day
```

Comment:
Same here, we have to use `yday`, no?

```suggestion
      if date.yday <= subscription_date.yday
```